### ISSUE TITLE
Rule proposal: no-implicit-coercion using a negative/minus (`-`) symbol

### ISSUE THREAD

**Issue by NotWoods** [typescript-eslint/typescript-eslint#7076]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

ESLint's [no-implicit-coercion](https://eslint.org/docs/latest/rules/no-implicit-coercion) rule can't check for casting a string to a number by using `-` because it would disallow negating a regular number. By using type information, we can provide additional coercion checks.

### Fail Cases

```typescript
const foo = '32px'
var n = -foo;
```

### Pass Cases

```typescript
const bar = 32
var n = -bar;
var n = -Number(foo);
var n = -parseFloat(foo);
var n = -parseInt(foo, 10);
```

**Comment by bradzacher** [comment]
This is a really interesting idea.
I'll be honest when I say that it's been a very long time since I've had to operate on numbers and needed to use the unary `-` to invert a sign.
I'd almost go as far as to ban it for non-literal numbers altogether (if you want to invert then use `* -1` which will be correctly checked by TS).

But this does seem like a niche, but useful check to have to ensure safety.

**Comment by NotWoods** [comment]
It might be worthwhile for the documentation for this rule to show how to configure [no-restricted-syntax](https://eslint.org/docs/latest/rules/no-restricted-syntax) to ban `-` altogether.

**Comment by samestep** [comment]
I've created an ESLint plugin with this rule: https://www.npmjs.com/package/eslint-plugin-unary-minus

**Comment by tylerlaprade** [comment]
Thank you @samestep, looks good! Any reason not to submit it as a PR to the main repo?

**Comment by samestep** [comment]
Thanks @tylerlaprade! The only reason was this conversation I had on Discord with @Zamiell yesterday:

> start by making your own custom rule yeah
> and then get everything working and write some tests, then you can submit a PR afterward

Should I submit a PR with this now? How can I do that?

**Comment by bradzacher** [comment]
Yeah there is no need to publish a separate package for a rule! Doing that involves a lot of extra boilerplate and effort - we don't recommend doing that unless you've got an idea for a rule that we haven't marked as accepted.

Even if you want to try the rule locally on your project before submitting a PR we suggest [local linking](https://typescript-eslint.io/contributing/local-development) for that to save you time!

**Comment by samestep** [comment]
Oh! Sorry about that, I guess I just misunderstood @Zamiell's message on Discord.
